pkg/services/cosmosdb/common-registered-provision.go

```diff
@@ -0,0 +1,178 @@
+	)
+}
+
+func (c *commonRegisteredManager) fillInCommonCredentials(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	databaseAccountName := instance.ProvisioningParameters.GetString("accountName") // nolint: lll
+	pk, err := getPrimaryKey(
+		ctx,
+		databaseAccountName,
+		c.databaseAccountsClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	dt := &cosmosdbInstanceDetails{}
+	dt.DatabaseAccountName = databaseAccountName
+	dt.PrimaryKey = pk
+	dt.FullyQualifiedDomainName = fmt.Sprintf(
+		"https://%s.documents.azure.com:443/",
+		instance.ProvisioningParameters.GetString("accountName"),
+	)
+	dt.ConnectionString = service.SecureString(
+		fmt.Sprintf("AccountEndpoint=%s;AccountKey=%s;",
+			dt.FullyQualifiedDomainName,
+			dt.PrimaryKey,
+		),
+	)
+	return dt, nil
+}
+
+func (c *commonRegisteredManager) fillInDifferentiatedCredentials(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	serviceName := instance.Service.GetName()
+	dt := instance.Details.(*cosmosdbInstanceDetails)
+
+	switch serviceName {
+	case "azure-cosmosdb-mongo-account-registered":
+		dt.ConnectionString = service.SecureString(
+			fmt.Sprintf(
+				"mongodb://%s:%s@%s:10255/?ssl=true&replicaSet=globaldb",
+				dt.DatabaseAccountName,
+				dt.PrimaryKey,
+				dt.FullyQualifiedDomainName,
+			),
+		)
+		// Allow to remove the https:// and the port 443 on the FQDN
+		// This will allow to adapt the FQDN for Azure Public / Azure Gov ...
+		// Before :
+		// https://6bd965fd-a916-4c3c-9606-161ec4d726bf.documents.azure.com:443
+		// After :
+		// 6bd965fd-a916-4c3c-9606-161ec4d726bf.documents.azure.com
+		hostnameNoHTTPS := strings.Join(
+			strings.Split(dt.FullyQualifiedDomainName, "https://"),
+			"",
+		)
+		dt.FullyQualifiedDomainName = strings.Join(
+			strings.Split(hostnameNoHTTPS, ":443/"),
+			"",
+		)
+	case "azure-cosmosdb-table-account-registered":
+		dt.ConnectionString = service.SecureString(
+			fmt.Sprintf(
+				"DefaultEndpointsProtocol=https;AccountName=%s;"+
+					"AccountKey=%s;TableEndpoint=%s",
+				dt.DatabaseAccountName,
+				dt.FullyQualifiedDomainName,
+				dt.PrimaryKey,
+			),
+		)
+	case "azure-cosmosdb-sql-account-registered":
+	case "azure-cosmosdb-graph-account-registered":
+		dt.ConnectionString = service.SecureString(
+			fmt.Sprintf("AccountEndpoint=%s;AccountKey=%s;",
+				dt.FullyQualifiedDomainName,
+				dt.PrimaryKey,
+			),
+		)
+	default:
+		return nil, fmt.Errorf("given service name %s is not vaild", serviceName)
+	}
+	return dt, nil
+}
+func getPrimaryKey(
+	ctx context.Context,
+	accountName string,
+	databaseAccountClient cosmosSDK.DatabaseAccountsClient,
+) (service.SecureString, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	allAccounts, err := databaseAccountClient.List(ctx)
```

Comment:
I know that we can find the account by only a name without resource group. So that less required provisioning parameters make it easier to use. But this operation is not always allowed for all kinds of service principals. A known typical scenario: the broker user has only a resource-group-scoped SP, then he doesn't have enough privilege to list all accounts in a subscription -- ARM would return 401/403 (I don't remember which it is). So, you should add resource group to provisioning parameters schema and just directly locate it here.

Reply:
How about make "resouceGroup" as an optional parameter. If provided, search will be executed insides the resource group; otherwise, will search the whole subscription.

Reply:
I don't think the flexibility here has enough value to increase the complexity of usage (doc) and code. Can we directly get the account by resource group and its name? https://github.com/Azure/azure-sdk-for-go/blob/master/services/cosmos-db/mgmt/2015-04-08/documentdb/databaseaccounts.go#L387

Reply:
The reason why I want to make it optional is to enable migration script. The CF api can't provide the provision parameter, and then we can't get the resource group of an instance in the script, which will disable the migration script. Without the script, the user must manually do migration one by one, even though their service principal is subscription scope.